File: server/models/goalModel.js
const mongoose = require('mongoose');

const goalSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User ID is required'],
    index: true
  },
  name: {
    type: String,
    required: [true, 'Goal name is required'],
    trim: true,
    maxlength: [50, 'Goal name cannot exceed 50 characters']
  },
  category: {
    type: String,
    required: [true, 'Goal category is required'],
    enum: ['Savings', 'Travel', 'Transportation', 'Technology', 'Emergency', 'Investment', 'Home', 'Education', 'Health', 'Other']
  },
  priority: {
    type: String,
    enum: ['High', 'Medium', 'Low'],
    default: 'Medium'
  },
  targetAmount: {
    type: Number,
    required: [true, 'Target amount is required'],
    min: [0, 'Target amount cannot be negative']
  },
  savedAmount: {
    type: Number,
    default: 0,
    min: [0, 'Saved amount cannot be negative']
  },
  targetDate: {
    type: Date,
    required: [true, 'Target date is required']
  },
  status: {
    type: String,
    enum: ['Active', 'Completed', 'Paused'],
    default: 'Active'
  },
  completedDate: {
    type: Date
  },
  description: {
    type: String,
    maxlength: [200, 'Description cannot exceed 200 characters']
  },
  isGuestMigrated: {
    type: Boolean,
    default: false
  },
  migratedAt: Date
}, {
  timestamps: true
});

// Fixed indexes to use 'userId' instead of 'profileId'
goalSchema.index({ userId: 1, status: 1 });
goalSchema.index({ userId: 1, targetDate: 1 });
goalSchema.index({ userId: 1, category: 1 });

// Virtual for progress percentage
goalSchema.virtual('progressPercentage').get(function() {
  if (this.targetAmount <= 0) return 0;
  return Math.round((this.savedAmount / this.targetAmount) * 100);
});

// Virtual for remaining amount
goalSchema.virtual('remainingAmount').get(function() {
  return Math.max(0, this.targetAmount - this.savedAmount);
});

// Virtual for days remaining
goalSchema.virtual('daysRemaining').get(function() {
  const today = new Date();
  const targetDate = new Date(this.targetDate);
  const diffTime = targetDate - today;
  const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
  return diffDays;
});

// Virtual for overdue status
goalSchema.virtual('isOverdue').get(function() {
  return this.daysRemaining < 0 && this.status !== 'Completed';
});

// Method to update saved amount
goalSchema.methods.updateSavedAmount = function(amount) {
  this.savedAmount += amount;
  if (this.savedAmount >= this.targetAmount && this.status !== 'Completed') {
    this.status = 'Completed';
    this.completedDate = new Date();
  }
  return this.save();
};

// Method to check if goal is achieved
goalSchema.methods.isAchieved = function() {
  return this.savedAmount >= this.targetAmount;
};

// Fixed static method to use 'userId' instead of 'profileId'
goalSchema.statics.getByProfile = function(userId, status = null) {
  const query = { userId };
  if (status) query.status = status;
  return this.find(query).sort({ targetDate: 1 });
};

// Fixed static method to use 'userId' instead of 'profileId'
goalSchema.statics.getSummaryByProfile = async function(userId) {
  const goals = await this.find({ userId });
  
  return {
    totalGoals: goals.length,
    activeGoals: goals.filter(g => g.status === 'Active').length,
    completedGoals: goals.filter(g => g.status === 'Completed').length,
    pausedGoals: goals.filter(g => g.status === 'Paused').length,
    totalTargetAmount: goals.reduce((sum, g) => sum + g.targetAmount, 0),
    totalSavedAmount: goals.reduce((sum, g) => sum + g.savedAmount, 0),
    overallProgress: goals.length > 0 ? 
      Math.round((goals.reduce((sum, g) => sum + g.savedAmount, 0) / 
                  goals.reduce((sum, g) => sum + g.targetAmount, 0)) * 100) : 0
  };
};

// Pre-save middleware to update completion status
goalSchema.pre('save', function(next) {
  if (this.savedAmount >= this.targetAmount && this.status !== 'Completed') {
    this.status = 'Completed';
    this.completedDate = new Date();
  }
  next();
});

module.exports = mongoose.model('Goal', goalSchema);

File: chatServer/app/models/goalModel.py
from pydantic import BaseModel, Field, ConfigDict, field_validator
from typing import Optional, Literal
from datetime import datetime
from bson import ObjectId

from app.models.base import PyObjectId  # single shared definition


class GoalModel(BaseModel):
    """Goal model matching the MongoDB schema."""

    id: Optional[str] = Field(None, alias="_id")
    userId: str
    name: str = Field(..., max_length=50)
    category: Literal[
        "Savings", "Travel", "Transportation", "Technology", "Emergency",
        "Investment", "Home", "Education", "Health", "Other"
    ]
    priority: Literal["High", "Medium", "Low"] = "Medium"
    targetAmount: float = Field(..., ge=0)
    savedAmount: float = Field(default=0, ge=0)
    targetDate: datetime
    status: Literal["Active", "Completed", "Paused"] = "Active"
    completedDate: Optional[datetime] = None
    description: Optional[str] = Field(None, max_length=200)
    isGuestMigrated: bool = False
    migratedAt: Optional[datetime] = None
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None
    v: Optional[int] = Field(None, alias="__v")

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_encoders={ObjectId: str},
    )

    @field_validator("userId", mode="before")
    @classmethod
    def coerce_objectid_to_str(cls, v):
        return str(v) if isinstance(v, ObjectId) else v


class GoalCreate(BaseModel):
    """Schema for creating a new goal."""

    userId: str
    name: str = Field(..., max_length=50)
    category: Literal[
        "Savings", "Travel", "Transportation", "Technology", "Emergency",
        "Investment", "Home", "Education", "Health", "Other"
    ]
    priority: Literal["High", "Medium", "Low"] = "Medium"
    targetAmount: float = Field(..., ge=0)
    savedAmount: float = Field(default=0, ge=0)
    targetDate: datetime
    description: Optional[str] = Field(None, max_length=200)

    model_config = ConfigDict(populate_by_name=True)


class GoalUpdate(BaseModel):
    """Schema for updating an existing goal."""

    name: Optional[str] = Field(None, max_length=50)
    category: Optional[Literal[
        "Savings", "Travel", "Transportation", "Technology", "Emergency",
        "Investment", "Home", "Education", "Health", "Other"
    ]] = None
    priority: Optional[Literal["High", "Medium", "Low"]] = None
    targetAmount: Optional[float] = Field(None, ge=0)
    savedAmount: Optional[float] = Field(None, ge=0)
    targetDate: Optional[datetime] = None
    status: Optional[Literal["Active", "Completed", "Paused"]] = None
    description: Optional[str] = Field(None, max_length=200)

    model_config = ConfigDict(populate_by_name=True)


class GoalSummary(BaseModel):
    """Summary statistics for goals."""

    totalGoals: int = 0
    activeGoals: int = 0
    completedGoals: int = 0
    pausedGoals: int = 0
    totalTargetAmount: float = 0
    totalSavedAmount: float = 0
    overallProgress: float = 0


class GoalResponse(BaseModel):
    """Enhanced goal response with server-calculated virtual fields."""

    id: str
    userId: str
    name: str
    category: str
    priority: str
    targetAmount: float
    savedAmount: float
    targetDate: datetime
    status: str
    completedDate: Optional[datetime] = None
    description: Optional[str] = None
    isGuestMigrated: bool
    migratedAt: Optional[datetime] = None
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None
    # Virtual / calculated fields
    progressPercentage: int
    remainingAmount: float
    daysRemaining: int
    isOverdue: bool

    model_config = ConfigDict(populate_by_name=True)